pkg/assembler/backends/inmem/pkg.go

```diff
@@ -246,92 +246,105 @@ func (p *pkgVersionNode) setPkgEquals(id uint32) { p.pkgEquals = append(p.pkgEqu
 
 // Ingest Package
 func (c *demoClient) IngestPackage(ctx context.Context, input model.PkgInputSpec) (*model.Package, error) {
-	return c.ingestPackage(ctx, input, true)
-}
-
-func (c *demoClient) ingestPackage(ctx context.Context, input model.PkgInputSpec, readOnly bool) (*model.Package, error) {
-	lock(&c.m, readOnly)
-	defer unlock(&c.m, readOnly)
-
+	c.m.RLock()
 	namespacesStruct, hasNamespace := c.packages[input.Type]
+	c.m.RUnlock()
 	if !hasNamespace {
-		namespacesStruct = &pkgNamespaceStruct{
-			id:         c.getNextID(),
-			typeKey:    input.Type,
-			namespaces: pkgNamespaceMap{},
+		c.m.Lock()
+		namespacesStruct, hasNamespace = c.packages[input.Type]
```

Comment:
is this duplicated code with line 250 above? Multiple instances of this...is this for a reason?

Reply:
The node could have been added between the release of the read mutex and aquiring the write mutex.

Reply:
ah I see.

Reply:
https://github.com/guacsec/guac/pull/646#issuecomment-1485923075
It is messy to do the locks/unlocks and checking twice etc. that is why I did the re-entry on all the others.